src/dspbase.jl

```diff
@@ -117,76 +117,315 @@ function deconv(b::StridedVector{T}, a::StridedVector{T}) where T
+            )
+            block_region = CartesianIndices(
+                NTuple{N, UnitRange{Int}}(edge_range)
+            )
+            for block_pos in block_region
+                # Figure out which portion of the input needs to be transformed
+                block_idx = convert(NTuple{N, Int}, block_pos)
+                data_offset = save_blocksize .* (block_idx .- 1)
+                pad_before = max.(0, sv .- data_offset .- 1)
+                data_ideal_stop = data_offset .+ save_blocksize
+                pad_after = max.(0, data_ideal_stop .- su)
+                data_region = CartesianIndices(
+                    UnitRange.(
+                        u_start .+ data_offset .- sv .+ pad_before .+ 1,
+                        u_start .+ data_ideal_stop .- pad_after .- 1
+                    )
+                )
+
+                # Convolve portion of input
+                _zeropad!(tdbuff, u, pad_before .+ 1, data_region)
+                os_conv_block!(tdbuff, fdbuff, filter_fd, p, ip)
+
+                # Save convolved result to output
+                block_out_stop = min.(
+                    out_start .+ data_offset .+ save_blocksize .- 1,
+                    out_stop
+                )
+                block_out_region = CartesianIndices(
+                    UnitRange.(out_start .+ data_offset, block_out_stop)
+                )
+                deficit = max.(0, pad_after .- sv .+ 1)
+                valid_buff_region = CartesianIndices(
+                    UnitRange.(sv, nffts .- deficit)
+                )
+                copyto!(out, block_out_region, tdbuff, valid_buff_region)
+            end
+        end
+    end
+end
+
+# Assumes u is larger, or same size, as v
+function _conv_kern_os!(out, u::AbstractArray{<:Any, N}, v, su, sv, sout, nffts) where N
+    u_start = first.(axes(u))
+    out_axes = axes(out)
+    out_start = first.(out_axes)
+    out_stop = last.(out_axes)
+    save_blocksize = min.(sout, nffts .- sv .+ 1)
+    nblocks = cld.(sout, save_blocksize)
+
+    # Pre-allocation
+    tdbuff, fdbuff, p, ip = os_prepare_conv(u, nffts)
+
+    # Transform the smaller filter
+    _zeropad!(tdbuff, v)
+    filter_fd = os_filter_transform!(tdbuff, p)
+    filter_fd .*= 1 / prod(nffts) # Normalize once for brfft
+
+    center_block_ranges = UnitRange.(2, nblocks .- 1)
+    edge_blocks = map(nblock -> nblock > 1 ? [1, nblock] : [1], nblocks)
+    all_dims = 1:N
+    edge_range = Vector{UnitRange}(undef, N)
+    for n_edges in all_dims
+        _conv_kern_os_edge!(
+            out,
+            tdbuff,
+            fdbuff,
+            edge_range,
+            p,
+            ip,
+            Val{n_edges}(),
+            u,
+            filter_fd,
+            center_block_ranges,
+            edge_blocks,
+            all_dims,
+            save_blocksize,
+            su,
+            sv,
+            nffts,
+            out_start,
+            out_stop,
+            u_start
+        )
```

Comment:
Could this be made a closure to avoid passing a gazillion parameters? (Of course, not doing dynamic dispatch here and just inlining that function would be even better, but I guess the type instability would be pretty bad for performance and so would be having `these_blocks` be a vector?)

Reply:
Yeah, the problem is `Val{n_edges}()`, which is obviously not type stable. Using simple integers instead of `Val` types causes the code generated by `subsets` to be less than great, and more importantly the output of `subsets`, `edge_dims`, would no longer a `Tuple`, which leads to type instability. I could play around with it a bit more to see if I can out a better solution. I hadn't though of using a closure. In my previous experience with closures, it ends up forcing to compiler to do a lot of boxing and unboxing of captured variables. I'll try that.

Reply:
You're also right with `these_blocks`, it's a big performance win for it to be a `Tuple`, because then the output of the subsequent `ProductIterator` is type stable.

Reply:
Usually variables that are only assigned to once should be safe to capture in a closure.

Reply:
I tried a closure with variables being modified passed as arguments, but it still substantially changed the output of `@code_warntype`, and had more type instability for some reason.

Reply:
I guess I also don't like closures with a bunch of closed-over variables from a stylistic perspective. For me, it's hard to spot the closed-over variables, hard to separate out the closure and figure out what's going on with it or use things like `@code_warntype` on closures. Basically, I'd rather have a huge parameter list than a huge set of captured variables.